src/Microsoft.Sbom.Parsers.Spdx30SbomParser/Utils/InternalMetadataProviderIdentityExtensions.cs

```diff
@@ -0,0 +1,190 @@
+        {
+            sha1Checksums.Add(checksumArray
+                .Where(c => c.Algorithm == HashAlgorithmName.SHA1)
+                .Select(c => c.ChecksumValue)
+                .FirstOrDefault());
+        }
+
+        var packageChecksumString = string.Join(string.Empty, sha1Checksums.OrderBy(s => s));
+#pragma warning disable CA5350 // Suppress Do Not Use Weak Cryptographic Algorithms as we use SHA1 intentionally
+        var sha1Hasher = SHA1.Create();
+#pragma warning restore CA5350
+        var hashByteArray = sha1Hasher.ComputeHash(Encoding.Default.GetBytes(packageChecksumString));
+
+        var packageVerificationCode = new PackageVerificationCode
+        {
+            Algorithm = Entities.Enums.HashAlgorithm.sha1,
+            HashValue = Convert.ToHexString(hashByteArray).Replace("-", string.Empty).ToLowerInvariant(),
+        };
+        packageVerificationCode.AddSpdxId();
+        return packageVerificationCode;
+    }
+
+    /// <summary>
+    /// Get the version of the package. This can be provided directly be the user, or
+    /// if not, we can try to generate one based on the build parameters.
+    ///
+    /// If we are unable to generate a package name, we throw an <see cref="ArgumentException"/>
+    /// exception.
+    /// </summary>
+    /// <returns>The string package version.</returns>
+    public static string GetPackageVersion(this IInternalMetadataProvider internalMetadataProvider)
+    {
+        if (internalMetadataProvider is null)
+        {
+            throw new ArgumentNullException(nameof(internalMetadataProvider));
+        }
+
+        // First check if the user provided a package version.
+        if (internalMetadataProvider.TryGetMetadata(MetadataKey.PackageVersion, out string packageVersion))
+        {
+            return packageVersion;
+        }
+
+        // If the build id is provided, use that as version.
+        if (internalMetadataProvider.TryGetMetadata(MetadataKey.Build_BuildId, out string buildId))
+        {
+            return buildId;
+        }
+
+        // Right now we don't have any better way to version the package. Throw an exception for the user to
+        // provide a package version.
+        throw new ArgumentException($"Unable to generate a package version based on provided parameters. " +
+                                    $"Please provide the package version in the 'PackageVersion' parameter.");
+    }
+
+    public static string GetPackageSupplier(this IInternalMetadataProvider internalMetadataProvider)
+    {
+        if (internalMetadataProvider is null)
+        {
+            throw new ArgumentNullException(nameof(internalMetadataProvider));
+        }
+
+        // First check if the user provided a package version.
+        if (internalMetadataProvider.TryGetMetadata(MetadataKey.PackageSupplier, out object packageSupplier))
+        {
+            return packageSupplier as string;
```

Comment:
What's the type of `packageSupplier`? If it's always a string, then we can strongly type it in the call to `TryGetMetadata`. If it's not a string, then the cast will fail and we'll return `null`. Is that ok?

Reply:
Yes both of these calls you commented on can be strongly typed since they should always be strings. Other calls in the old 2.2 generator left it as an object type, which is why I retained it here. But I agree strong typing makes more sense here.